server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java

```diff
@@ -0,0 +1,179 @@
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.transport.mailets;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import javax.mail.MessagingException;
+
+import org.apache.james.protocols.lib.PortUtil;
+import org.apache.james.smtpserver.mock.util.MockSpamd;
+import org.apache.james.util.scanner.SpamAssassinInvoker;
+import org.apache.mailet.Mail;
+import org.apache.mailet.base.test.FakeMail;
+import org.apache.mailet.base.test.FakeMailetConfig;
+import org.apache.mailet.base.test.MimeMessageBuilder;
+import org.junit.Test;
+
+public class SpamAssassinTest {
+    private SpamAssassin mailet = new SpamAssassin();
+
+    @Test
+    public void initShouldNotThrowWhenNoSpamHost() throws Exception {
+        assertThatThrownBy(() -> mailet.init()).isInstanceOf(NullPointerException.class) ;
+    }
+
+    @Test
+    public void initShouldSetDefaultSpamdHostWhenDefault() throws Exception {
+        mailet.init(FakeMailetConfig.builder()
+            .mailetName("SpamAssassin")
+            .build());
+
+        assertThat(mailet.getSpamdHost()).isEqualTo(SpamAssassin.DEFAULT_HOST);
+    }
+
+    @Test
+    public void initShouldSetSpamdHostWhenPresent() throws Exception {
+        String spamdHost = "any.host";
+        mailet.init(FakeMailetConfig.builder()
+            .mailetName("SpamAssassin")
+            .setProperty(SpamAssassin.SPAMD_HOST, spamdHost)
+            .build());
+
+        assertThat(mailet.getSpamdHost()).isEqualTo(spamdHost);
+    }
+
+    @Test
+    public void initShouldSetDefaultSpamdPortWhenDefault() throws Exception {
+        mailet.init(FakeMailetConfig.builder()
+            .mailetName("SpamAssassin")
+            .build());
+
+        assertThat(mailet.getSpamdPort()).isEqualTo(SpamAssassin.DEFAULT_PORT);
+    }
+
+    @Test
+    public void initShouldThrowWhenSpamdPortIsNotNumber() throws Exception {
+        assertThatThrownBy(() -> mailet.init(FakeMailetConfig.builder()
+            .mailetName("SpamAssassin")
+            .setProperty(SpamAssassin.SPAMD_PORT, "noNumber")
+            .build())).isInstanceOf(MessagingException.class);
+    }
+
+    @Test
+    public void initShouldSetSpamPortWhenPresent() throws Exception {
+        int spamPort = 1000;
+        mailet.init(FakeMailetConfig.builder()
+            .mailetName("SpamAssassin")
+            .setProperty(SpamAssassin.SPAMD_PORT, "" + spamPort)
+            .build());
+
+        assertThat(mailet.getSpamdPort()).isEqualTo(spamPort);
+    }
+
+    @Test
+    public void serviceShouldWriteSpamAttributeOnMail() throws Exception {
+        int port = PortUtil.getNonPrivilegedPort();
+        new Thread(new MockSpamd(port)).start();
```

Comment:
If you start it, then we need to stop it at the end of your test. What about starting it / stopping it using @Before @After ?

And wrap your thread in an Executor (fixed threadpool something?)

Reply:
Pro tip:
 - Pass port 0 to get a random port
 - Add a getPort method to MockSpamD

Reply:
Would be even better to write a JUNIT rule for that no ?

Reply:
https://github.com/linagora/james-project/pull/1194/commits/ac79e58b211c7ad8641a72c02d98e8b561afda33 

```
JAMES-2262 Use a Test rule for Mock SpamD test rule 
```

Come on!